src/api/heap_profiler.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::{HashMap, VecDeque};
+use std::fs::File;
+use std::io::Write;
+use std::time::{Duration, SystemTime};
+use uptown_funk::host_functions;
+
+use log::debug;
+use log::error;
+
+type Ptr = u32;
+type Size = u32;
+const HISTORY_CAPACITY: usize = 100_000;
+
+pub struct HeapProfilerState {
+    memory: HashMap<Ptr, Size>,
+    started: SystemTime,
+    heap_history: VecDeque<(i32, Duration)>,
+}
+
+impl HeapProfilerState {
+    pub fn new() -> Self {
+        Self {
+            memory: HashMap::new(),
+            started: SystemTime::now(),
+            heap_history: VecDeque::new(),
+        }
+    }
+
+    // Merge child process profile results into parent process.
+    //
+    // Usually this should be called when child process exits.
+    pub fn merge(&mut self, profiler: HeapProfilerState) {
+        self.started = std::cmp::min(self.started, profiler.started);
+        let started_delta = profiler.started.duration_since(self.started).unwrap();
+        // TODO sorted state merging of two sorted lists can be done in O(N+M)
+        // currently this runs in O((N+M)*log(N+M)*C)
+        profiler.heap_history.iter().for_each(|(h, d)| {
+            self.heap_history.push_back((*h, *d + started_delta));
+        });
+        self.heap_history
+            .make_contiguous()
+            .sort_unstable_by_key(|&(_, d)| d);
+        // remove oldest extra elements (ringbuffer)
+        if self.heap_history.len() > HISTORY_CAPACITY {
+            self.heap_history
+                .drain(0..(self.heap_history.len() - HISTORY_CAPACITY));
+        }
+    }
+
+    // Deallocate remaining memory. This is invoked when process exits
+    pub fn free_all(&mut self) {
+        // implicitly deallocate all left over child memory
+        self.memory.clone().iter().for_each(|(k, _)| {
```

Comment:
`free_all` is simple (sum all values in the map and clear the map). You don't need to reuse `free_profiler` (and therefore no need to clone the map), just implement it directly. Something like this:
```rust
        self.history_push(-(self.memory.values().sum::<u32>() as i32));
        self.memory.clear();
```
Do I miss something?

Reply:
Yes, I think that should avoid the clone. ty

I wasn't aware of `sum` :| , and didn't want to reimplement `free_profiler` again so I decided to clone and reuse `free_profiler`for readability.